### PR TITLE
Add support for adding `cache-control` and `Postman-Token` headers

### PR DESCRIPTION
**Pull request by ankitsiam**

### CODE REVIEW

File: lib/requester/core.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 var _ = require('lodash'),
     dns = require('dns'),
     Socket = require('net').Socket,
+    uuid = require('uuid').v4,
```
Comment (codenirvana) [inline lib/requester/core.js]:
```suggestion
    uuid = require('uuid/v4'),
```

File: lib/requester/core.js [hunk 2]
```diff
@@ -67,7 +68,14 @@ var _ = require('lodash'),
 
         // removes the `referer` header when a redirect happens (default: false)
         // @note `referer` header set in the initial request will be preserved during redirect chain
-        removeRefererHeader: 'removeRefererHeaderOnRedirect'
+        removeRefererHeader: 'removeRefererHeaderOnRedirect',
+
+        // Controls whether or not request should be sent with cache-control header, if true request will contain
+        // 'cache-control: no-cache' header
+        noCacheHeader: 'noCacheHeader',
```
Comment (codenirvana) [inline lib/requester/core.js]: Why are these added to protocol profile behaviors?

File: lib/requester/core.js [hunk 3]
```diff
@@ -259,8 +269,16 @@ module.exports = {
         self.ensureHeaderExists(options.headers, 'User-Agent', 'PostmanRuntime/' + version);
         self.ensureHeaderExists(options.headers, 'Accept', '*/*');
 
+
+        (_.isUndefined(options.noCacheHeader) ? true : options.noCacheHeader) &&
```
Comment (codenirvana) [inline lib/requester/core.js]: Just `options.noCacheHeader && ...`

File: lib/requester/core.js [hunk 4]
```diff
@@ -259,8 +269,16 @@ module.exports = {
...
+            self.ensureHeaderExists(options.headers, 'Cache-Control', 'no-cache');
+
+        (_.isUndefined(options.postmanTokenHeader) ? true : options.postmanTokenHeader) &&
+            self.ensureHeaderExists(options.headers, 'Postman-Token', uuid.v4());
+
         // The underlying Node client does add the host header by itself, but we add it anyway, so that
         // it is bubbled up to us after the request is made. If left to the underlying core, it's not :/
+
```
Comment (codenirvana) [inline lib/requester/core.js]: Extra new line

File: test/integration/protocol-profile-behavior/sanity.test.js [hunk 5]
```diff
@@ -368,6 +368,281 @@ describe('protocolProfileBehavior', function () {
         });
     });
 
+    describe('with noCacheHeader: true', function () {
```
Comment (codenirvana) [inline test/integration/protocol-profile-behavior/sanity.test.js]: these tests are added in the wrong test suite.
Also, why are these tested with redirects?

File: test/integration/request-flow/headers.test.js [hunk 6]
```diff
@@ -82,6 +82,25 @@ describe('request headers', function () {
                                 value: 'value3'
                             }]
                         }
+                    },
```
Comment (codenirvana) [inline test/integration/request-flow/headers.test.js]: Again, wrong place. These belong `requester-spec` suite.

File: lib/requester/core.js (hunk 2 above)
Comment (ankitsiam) [reply lib/requester/core.js]: Coz you would want to override the behaviour for maybe a few individual requests and not for the entire collection ?

File: lib/requester/core.js (hunk 3 above)
Comment (ankitsiam) [reply lib/requester/core.js]: We also want to send these two headers by default, if requester options or PPB options are not provided to override this behaviour.

Comment (codenirvana) [reply lib/requester/core.js]: How can one set `noCacheHeader`? You need to first add it to requester options.

Comment (ankitsiam) [reply lib/requester/core.js]: Requester options, PPB or custom cache-control header, but if none is provided by default we want to send these two, unless somebody overrides them with custom headers or disables them with options.

Comment (codenirvana) [reply lib/requester/core.js]: Right, but here: https://github.com/postmanlabs/postman-runtime/pull/740/files#diff-4a97c1387dbab5c3fb9f31a111b03b56R378

I see `noCacheHeader: false` but assertion checks for its existence.

Comment (ankitsiam) [reply lib/requester/core.js]: No it's correct, we are overriding the requester option by PBB, where we set it to true.

File: lib/requester/core.js [hunk 7]
```diff
@@ -235,6 +238,8 @@ module.exports = {
         options.timeout = defaultOpts.timeout;
         options.gzip = true;
         options.time = defaultOpts.timings;
+        options.noCacheHeader = defaultOpts.noCacheHeader;
```
Comment (codenirvana) [inline lib/requester/core.js]: these are not `postman-request` options so this can be avoided.

File: lib/requester/requester-pool.js [hunk 8]
```diff
@@ -16,6 +16,8 @@ RequesterPool = function (options) {
         followRedirects: _.get(options, 'requester.followRedirects', true),
         followOriginalHttpMethod: _.get(options, 'requester.followOriginalHttpMethod'),
         maxRedirects: _.get(options, 'requester.maxRedirects'),
+        noCacheHeader: _.get(options, 'requester.noCacheHeader', true),
```
Comment (codenirvana) [inline lib/requester/requester-pool.js]: consider using `cacheControlHeader`, `noCacheHeader` is confusing.

File: test/integration/requester-spec/headers.test.js [hunk 9]
```diff
@@ -0,0 +1,236 @@
+var _ = require('lodash'),
+    sinon = require('sinon'),
+    expect = require('chai').expect;
+
+describe('cache-control and postman-token headers', function () {
+    var testrun,
+        HOST = 'https://www.postman-echo.com/get';
+
+    describe('when `noCache` and `postmanToken` requester options are not provided', function() {
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: I don't feel `describe` should have a description like this. 😕

File: lib/requester/core.js [hunk 10]
```diff
@@ -259,6 +262,9 @@ module.exports = {
         self.ensureHeaderExists(options.headers, 'User-Agent', 'PostmanRuntime/' + version);
         self.ensureHeaderExists(options.headers, 'Accept', '*/*');
 
+        defaultOpts.cacheControlHeader && self.ensureHeaderExists(options.headers, 'Cache-Control', 'no-cache');
+        defaultOpts.postmanTokenHeader && self.ensureHeaderExists(options.headers, 'Postman-Token', uuid.v4());
```
Comment (codenirvana) [inline lib/requester/core.js]:
```suggestion
        defaultOpts.postmanTokenHeader && self.ensureHeaderExists(options.headers, 'Postman-Token', uuid());
```

File: test/integration/requester-spec/headers.test.js [hunk 11]
```diff
@@ -0,0 +1,236 @@
+var _ = require('lodash'),
+    sinon = require('sinon'),
+    expect = require('chai').expect;
+
+describe('cache-control and postman-token headers', function () {
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]:
```suggestion
describe('Requester Spec: cacheControlHeader and postmanTokenHeader', function () {
```

File: test/integration/requester-spec/headers.test.js [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    var testrun,
+        HOST = 'https://www.postman-echo.com/get';
+
+    describe('are not provided', function() {
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]:
```suggestion
    describe('with undefined', function() {
```

File: test/integration/requester-spec/headers.test.js [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+            expect(request.header).to.deep.include({key: 'Cache-Control', value: 'no-cache'});
+            expect(response.headers).to.have.property('cache-control', 'no-cache');
+        });
+
+        it('should send request with `Postman-Token` header', function () {
+            var request = testrun.request.getCall(0).args[3].toJSON(),
+                response = testrun.response.getCall(0).args[2].stream.toString(),
+
+                regex = /^(Postman-Token: )?[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/mi,
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: just keep uuid-v4 regex part.

File: test/integration/requester-spec/headers.test.js [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+                responseHeader = _.get(JSON.parse(response).headers, 'postman-token'),
+
+                requestHeaders = request.header.filter(function(header) {
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: why `filter` required here?

File: test/integration/requester-spec/headers.test.js [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+                    return (header.key === 'Postman-Token');
+                });
+
+            expect(regex.test(responseHeader)).to.be.true;
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: check for existing of the property(like its done in the first test) and later assert the value to be valid uuid-v4.

File: test/integration/requester-spec/headers.test.js [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+            expect(regex.test(requestHeaders[0].value)).to.be.true;
+        });
+    });
+
+    describe('are set to true', function() {
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]:
```suggestion
    describe('with true', function() {
```

File: test/integration/requester-spec/headers.test.js [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+                requestHeaders = request.header.filter(function(header) {
+                    return (header.key === 'Postman-Token');
+                });
+
+            expect(regex.test(responseHeader)).to.be.true;
+            expect(regex.test(requestHeaders[0].value)).to.be.true;
+        });
+    });
+
+    describe('are set to false', function() {
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]:
```suggestion
    describe('with false', function() {
```

File: test/integration/requester-spec/headers.test.js [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+                requestHeaders = request.header.filter(function(header) {
+                    return (header.key === 'Postman-Token');
+                });
+
+            expect(regex.test(responseHeader)).to.be.false;
+            expect(requestHeaders).to.be.empty;
+        });
+    });
+
+    describe('custom headers are provided', function() {
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]:
```suggestion
    describe('with custom headers', function() {
```

File: test/integration/requester-spec/headers.test.js [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+        before(function(done) {
+            this.run({
+                requester: {
+                    cacheControlHeader: false,
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: these should be true to make sure it will not overwrite if true.

Comment (ankitsiam) [reply test/integration/requester-spec/headers.test.js]: shouldn't they be overwritten irrespective of true or false? since headers are in a request this would mean manually removing headers if options are true? I feel by default custom headers should have higher priority?

File: test/integration/requester-spec/headers.test.js [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+            expect(response.headers).to.have.property('cache-control', 'no-cache');
+        });
+
+        it('should send request with `Postman-Token` header', function () {
+            var request = testrun.request.getCall(0).args[3].toJSON(),
+                response = testrun.response.getCall(0).args[2].stream.toString(),
+                responseHeader = _.get(JSON.parse(response).headers, 'postman-token'),
+                regex = /[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/mi;
+
+            expect(regex.test(responseHeader)).to.be.true;
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]:
```suggestion
            expect(response.headers).to.have.property('postman-token').that. match(uuidRegex);
```

File: test/integration/requester-spec/headers.test.js [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+
+            expect(request.header).to.deep.include({key: 'Cache-Control', value: 'no-cache'});
+            expect(response.headers).to.have.property('cache-control', 'no-cache');
+        });
+
+        it('should send request with `Postman-Token` header', function () {
+            var request = testrun.request.getCall(0).args[3].toJSON(),
+                response = testrun.response.getCall(0).args[2].stream.toString(),
+                regex = /[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/mi,
+                responseHeader = _.get(JSON.parse(response).headers, 'postman-token');
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: same as above.

File: test/integration/requester-spec/headers.test.js [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+            sinon.assert.calledOnce(testrun.response);
+            sinon.assert.calledWith(testrun.response.getCall(0), null);
+        });
+
+        it('should send request without `Cache-Control: no-cache` header', function () {
+            var request = testrun.request.getCall(0).args[3].toJSON(),
+                response = JSON.parse(testrun.response.getCall(0).args[2].stream.toString());
+
+            expect(request.header).to.deep.not.include({key: 'Cache-Control', value: 'no-cache'});
+            expect(response.headers).to.not.have.property('cache-control', 'no-cache');
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]:
```suggestion
            expect(response.headers).to.not.have.property('cache-control');
```

File: test/integration/requester-spec/headers.test.js [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+
+            sinon.assert.calledOnce(testrun.response);
+            sinon.assert.calledWith(testrun.response.getCall(0), null);
+        });
+
+        it('should send request without `Cache-Control: no-cache` header', function () {
+            var request = testrun.request.getCall(0).args[3].toJSON(),
+                response = JSON.parse(testrun.response.getCall(0).args[2].stream.toString());
+
+            expect(request.header).to.deep.not.include({key: 'Cache-Control', value: 'no-cache'});
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: just check for missing key, don't check for the exact value.

File: test/integration/requester-spec/headers.test.js [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+            expect(response.headers).to.not.have.property('cache-control', 'no-cache');
+        });
+
+        it('should send request without `Postman-Token` header', function () {
+            var request = testrun.request.getCall(0).args[3].toJSON(),
+                response = testrun.response.getCall(0).args[2].stream.toString(),
+                regex = /[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/mi,
+                responseHeader = _.get(JSON.parse(response).headers, 'postman-token');
+
+            expect(regex.test(responseHeader)).to.be.false;
```
Comment (codenirvana) [inline test/integration/requester-spec/headers.test.js]: this is not required here, just check for missing key.

File: lib/requester/core.js [hunk 25]
```diff
@@ -259,6 +262,9 @@ module.exports = {
         self.ensureHeaderExists(options.headers, 'User-Agent', 'PostmanRuntime/' + version);
         self.ensureHeaderExists(options.headers, 'Accept', '*/*');
 
+        defaultOpts.implicitCacheControl && self.ensureHeaderExists(options.headers, 'Cache-Control', 'no-cache');
+        defaultOpts.implicitTraceHeader && self.ensureHeaderExists(options.headers, 'Postman-Token', uuid());
```
Comment (kunagpal) [inline lib/requester/core.js]: `uuid` will still get invoked when a trace header already exists.

Comment (kunagpal) [reply lib/requester/core.js]: Add a TODO for this.
